3-Solving-Problems-By-Searching/bi-directional.js

```diff
@@ -0,0 +1,139 @@
+//Function to clone a object
+function cloneObject(obj) {
+  return JSON.parse(JSON.stringify(obj));
+}
+
+function distance(point1, point2) {
+  return Math.sqrt(Math.pow(point1[0] - point2[0], 2) + Math.pow(point1[1] - point2[1], 2));
+}
+class Node {
+  constructor(id, x, y, adjacent) {
+    this.id = id;
+    this.x = x;
+    this.y = y;
+    this.adjacent = (adjacent != undefined) ? adjacent : [];
+  }
+}
+
+function randomPlanarGraph(height, width, radius) {
+  let [points, grid] = poissonDistribution(height, width, radius);
+
+  let nodes = [];
+  let edges = [];
+  for (let i = 0; i < points.length; i++) {
+    nodes.push(new Node(i, points[i][0], points[i][1]));
+  }
+
+  for (let i = 0; i < nodes.length; i++) {
+    for (let j = i + 1; j < nodes.length; j++) {
+      if (distance(points[i], points[j]) < 2 * radius) {
+        if (Math.random() < 0.9) {
+          nodes[i].adjacent.push(j);
+          nodes[j].adjacent.push(i);
+          edges.push([i, j]);
+        }
+      }
+    }
+  }
+  return [nodes, edges];
+}
+class Graph {
+  constructor(height, width, radius) {
+    this.totalNodes = 20;
+    this.nodes = [];
+    this.height = height;
+    this.width = width;
+    this.radius = radius;
+    [this.nodes, this.edges] = randomPlanarGraph(this.height, this.width, this.radius);
+  }
+
+  getAdjacent(id) {
+    return this.nodes[id].adjacent;
+  }
+}
+
+class BreadthFirstSearch {
```

Comment:
The earlier Breadth First Search was simply a 3 line code that takes the problem object and returns the first element from its frontier. Here, the Breadth First Search class performs what the class 'GraphAgent' performed earlier. But in this particular case, I am using generators. Generators prove to be especially useful here since there are 2 bfs running simultaneously and then there is one diagram that extracts nodes from the envelope iterate() function from Bi-directional Problem.
Additionally, the implementation of the graphs is also different. For example, the getAdjacent function earlier simply scanned the edges list to get the nodes adjacent to a node; but in this case, that will be too costly since there are thousands of nodes. So here, we keep a list of adjacent nodes in the graph node itself. Similarly, there are more fundamental differences between the implementations. Due to all this reasons, I thought trying to reuse the code would get in my way of making the diagram.
What are your thoughts on this part?